### ISSUE TITLE
RenderingServer.global_shader_parameter_add doesn't seem to add a parameter

### ISSUE THREAD

**Issue by adamscoble** [godotengine/godot#77988]
### Godot version

4.1.dev4.official

### System information

Godot v4.1.dev4 - Windows 10.0.19045 - Vulkan (Mobile) - dedicated NVIDIA GeForce GTX 1080 (NVIDIA; 31.0.15.3141) - Intel(R) Core(TM) i7-4770 CPU @ 3.40GHz (8 Threads)

### Issue description

Using `RenderingServer.global_shader_parameter_add` doesn't appear to add an entry in Project Settings -> Shader Globals, and the shader editor doesn't pick up on its existence. 

### Steps to reproduce

Here is an example script to place on a SubViewport. Add it to a scene and note the global is not added.

```
@tool
extends SubViewport


func _ready() -> void:
	await RenderingServer.frame_post_draw
	RenderingServer.global_shader_parameter_add(HEIGHT_MAP_GLOBAL, RenderingServer.GLOBAL_VAR_TYPE_SAMPLER2D, get_texture())
```

### Minimal reproduction project

See above

**Comment by adamscoble** [comment]
In case it's relevant, I just tried to add it view the editor and it told me it already exists, then the editor crashed.

**Comment by AThousandShips** [comment]
I don't believe this is supported, see [here](https://docs.godotengine.org/en/stable/tutorials/shaders/shader_reference/shading_language.html#global-uniforms)

**Comment by Calinou** [comment]
It should be possible for scripts to manipulate shader globals in a way that is visible in the Project Settings, so that editor plugins can create/modify/remove shader globals.

**Comment by jsjtxietian** [comment]
> But would be good to add functionality to script side, unsure how to access it via `ProjectSettings`

What about add functions like `add_shader_global` to ProjectSettings? The setting itself can be acquired by `get_setting` but we still need to call `global_shader_parameter_add`, so add a new function somewhere is needed.

**Comment by noidexe** [comment]
In 4.2.1 I can confirm doing RenderingServer.global_shader_parameter_add() in a tool script will make the parameters show up in ProjectSettings.

However, they are not really there when the game runs.
- RenderingServer will still complain about the missing global uniforms
- if you do ProjectSettings.get_property_list() you won't find them listed there
- Doing ProjectSettings.save() after RenderingServer.global_shader_parameter_add() will trigger the popup saying that project.godot is newer on disk, but if you check on git the file has not been changed at all

I have found a workaround. Instead of using RenderingServer.global_shader_parameter_add() use this: 

```gdscript
func _add_shader_global(param_name: StringName, type: RenderingServer.GlobalShaderParameterType, default_value: Variant, overwrite := false) -> void:
	const prefix = "shader_globals/"

	# from https://github.com/godotengine/godot/blob/17e7f85c06366b427e5068c5b3e2940e27ff5f1d/servers/rendering/renderer_rd/storage_rd/material_storage.cpp#L1614
	const enum_to_string = ["bool","bvec2","bvec3","bvec4","int","ivec2","ivec3","ivec4","rect2i",
							"uint","uvec2","uvec3","uvec4","float","vec2","vec3","vec4","color",
							"rect2","mat2","mat3","mat4","transform_2d","transform","sampler2D",
							"sampler2DArray","sampler3D","samplerCube"]

	# only the path is serialized for those
	const resource_types = ["sampler2D","sampler2DArray","sampler3D","samplerCube"]
	
	var exists = ProjectSettings.has_setting(prefix + param_name)

	if not overwrite and exists:
		return
	
	var type_as_string = enum_to_string[type]
	if type_as_string in resource_types and default_value is Resource:
		default_value = (default_value as Resource).resource_path
	
	# global shader params are saved in project.godot as a dictionary with 
	# the following structure
	var dict = {
		"type" = type_as_string,
		"value" = default_value
	}
	
	# Save the param to disk
	ProjectSettings.set_setting(prefix + param_name, dict)
	ProjectSettings.save()
	
	# We still need to manually add them to the RenderingServer so they are available
	# in the editor
	if exists:
		RenderingServer.global_shader_parameter_set(param_name, default_value)
	else:
		RenderingServer.global_shader_parameter_add(param_name, type, default_value)
```
You should call it as early as possible so the parameters are available in the editor to any open scenes with meshes that use them. 

However we definitely need a built in way to do this

**Comment by clayjohn** [comment]
To add some context here, it seems the design of global shader parameters did not consider adding permanent shader parameters at run time (either through a tool script or running the game). The ProjectSettings manages its global shader paramaters essentially the same way that @noidexe has proposed in the post above. 

I agree we need to have a better way to manage this as plugins are going to need to rely on global shader paramaters and it sucks if users have to register the global paramaters in the ProjectSettings themselves. 

Unless someone else disagrees, I think we should implement @jsjtxietian's suggestion and add a helper function to ProjectSettings that allows you to register a global shader parameter.

**Comment by noidexe** [comment]
> it sucks if users have to register the global paramaters in the ProjectSettings themselves.

Also if you download any assets that rely on global shader params you need to make sure you have all the params present in Project Settings before you open any of those assets in the editor. Otherwise shader compilation will fail. If shader compilation fails (local) shader parameters are no longer exposed to the material editor and as soon as the material gets saved they are gone from disk and memory. 

![image](https://github.com/godotengine/godot/assets/526829/c960c292-9cb3-423a-afa6-1285ad92a4e6)

**Comment by xuc1995** [comment]
> 在 4.2.1 中，我可以确认在工具脚本中执行 RenderingServer.global_shader_parameter_add（） 将使参数显示在 ProjectSettings 中。
> 
> 然而，当游戏运行时，他们并不真正在那里。
> 
> * RenderingServer 仍然会抱怨缺少全局 uniform
> * 如果你这样做 ProjectSettings.get_property_list（） 你不会在那里找到它们
> * 在 RenderingServer.global_shader_parameter_add（） 之后执行 ProjectSettings.save（） 会触发弹出窗口，指出 project.godot 在磁盘上较新，但如果你在 git 上检查，文件根本没有被更改
> 
> 我找到了一个解决方法。不要使用 RenderingServer.global_shader_parameter_add（），请使用以下内容：
> 
> ```gdscript
> func _add_shader_global(param_name: StringName, type: RenderingServer.GlobalShaderParameterType, default_value, overwrite := false) -> void:
> 	const prefix = "shader_globals/"
> 
> 	# from https://github.com/godotengine/godot/blob/17e7f85c06366b427e5068c5b3e2940e27ff5f1d/servers/rendering/renderer_rd/storage_rd/material_storage.cpp#L1614
> 	const enum_to_string = ["bool","bvec2","bvec3","bvec4","int","ivec2","ivec3","ivec4","rect2i",
> 							"uint","uvec2","uvec3","uvec4","float","vec2","vec3","vec4","color",
> 							"rect2","mat2","mat3","mat4","transform_2d","transform","sampler2D",
> 							"sampler2DArray","sampler3D","samplerCube"]
> 
> 	# only the path is serialized for those
> 	const resource_types = ["sampler2D","sampler2DArray","sampler3D","samplerCube"]
> 	
> 	var exists = ProjectSettings.has_setting(prefix + param_name)
> 
> 	if not overwrite and exists:
> 		return
> 	
> 	var type_as_string = enum_to_string[type]
> 	var value = default_value
> 	
> 	if type_as_string in resource_types and default_value is Resource:
> 		value = (default_value as Resource).resource_path
> 	
> 	# global shader params are saved in project.godot as a dictionary with 
> 	# the following structure
> 	var dict = {
> 		"type" = type_as_string,
> 		"value" = value
> 	}
> 	
> 	# Save the param to disk
> 	ProjectSettings.set_setting(prefix + param_name, dict)
> 	ProjectSettings.save()
> 	
> 	# We still need to manually add them to the RenderingServer so they are available
> 	# in the editor
> 	if exists:
> 		RenderingServer.global_shader_parameter_set(param_name, default_value)
> 	else:
> 		RenderingServer.global_shader_parameter_add(param_name, type, default_value)
> ```
> 
> 您应该尽早调用它，以便参数在编辑器中可用于任何具有使用它们的网格的打开场景。
> 
> 但是，我们肯定需要一个内置的方法来做到这一点

![image](https://github.com/user-attachments/assets/d12cfc1f-9de7-4b54-be79-357d1ac9d496)
Thanks a lot, it do works for me.

**Comment by VP-GAMES** [comment]
I have the same issue in godot4.4 (Ubuntu 24.10)
I created some unit tests with gdUnit4 for basic data types => bool, int, float, vec2, vec3[globalshaderparameter.zip](https://github.com/user-attachments/files/18353601/globalshaderparameter.zip)

![Image](https://github.com/user-attachments/assets/7c35090d-abd5-48c2-8dd0-09395427a16a)
I hope it helps to fix the bug!
